### PR TITLE
iMX6: repackage cubox/udoo/wandboard images to use FDTDIR

### PR DESCRIPTION
**Pull request by chewitt**
This combines https://github.com/by-gnome/LibreELEC.tv/commit/de7b1d38ec57a71854afe4eb948e5ea32848578d and #4917 testing with FDTDIR to halve the number of iMX6 images from six to three. I've also added `canupdate.sh` to block upgrades from legacy images; total user numbers are small and support will be easier if we force a clean install. If testing subsequently proves upgrading LE 8.0/8.2 > LE 10.0 can be done reliably we can always merge another PR. Until then, it's been years since iMX6 had an active maintainer so let's keep it simple.

@by-gnome @etrax79 please test the images here https://chewitt.libreelec.tv/testing/9.80/ as I don't have iMX6 h/w

### CODE REVIEW

**Comment by lrusak** [comment]
I would really rather this be generic for all projects. I'm also not sure why this was PR'd over #4917?

The entire point of uboot_helper is to define what we want to use for each board. Using wildcards negates this.

**Comment by chewitt** [comment]
Amlogic uses wildcards: https://github.com/LibreELEC/LibreELEC.tv/blob/master/projects/Amlogic/bootloader/release#L16

Allwinner uses wildcards: https://github.com/LibreELEC/LibreELEC.tv/blob/master/projects/Allwinner/bootloader/release#L12

Eynox uses wildcards: https://github.com/LibreELEC/LibreELEC.tv/blob/master/projects/Samsung/bootloader/release#L8

RK uses wildcards: https://github.com/LibreELEC/LibreELEC.tv/blob/master/projects/Samsung/bootloader/release#L8

The sole difference is that I'm using them in `mkimage` not `releases` so the tar file contains the full selection of iMX6 dtb files (should people want to experiment with boards beyond the limited selection we historically shipped) but the outcome in the root folder of the .img file is the same; only the boards we historically supported.

**Comment by etrax79** [comment]
@chewitt I will download and try both, installation and update. But as your result is nearly the same I would not expect any problems.
The wildcard-less way for uboot_helper and mkimage was also prepared (in a new branch) and linked in the last comment on #4917 . Please check, discuss and maybe we could find a merged solution which satisfies everybody... Feel free to copy and paste, if you like

**Comment by lrusak** [comment]
You are confusing two different things here:

We copy all dtbs to /usr/share/bootloader here: https://github.com/LibreELEC/LibreELEC.tv/blob/7b30b3d50515d0b6ff60db43df94474abe3ca6c4/packages/linux/package.mk#L268-L272

This is done so that they can be copied to the release tarball. This is needed so that we can ship a common project/device tarball that will cover all the boards listed in `uboot_helper`. This is done in the project specific release script which as you pointed out is a wildcard glob. see: https://github.com/LibreELEC/LibreELEC.tv/blob/7b30b3d50515d0b6ff60db43df94474abe3ca6c4/projects/NXP/devices/iMX6/bootloader/release#L4-L5
and
https://github.com/LibreELEC/LibreELEC.tv/blob/7b30b3d50515d0b6ff60db43df94474abe3ca6c4/projects/NXP/devices/iMX6/bootloader/release#L13

This could probably be made more generic as we basically do the same thing for all projects/devices.

We use the `dtb` option in `uboot_helper` to define the dtb file that we want to have installed for the specific `.img` file that is created for the specific board that is listed. See: https://github.com/LibreELEC/LibreELEC.tv/blob/7b30b3d50515d0b6ff60db43df94474abe3ca6c4/scripts/mkimage#L215-L220


In the case of i.MX6 there is some special u-boot logic that looks for different dtbs based on the base system (as you can see cubox has q and dl variants). So if we want to have both supported we should list both dtb's in the `uboot_helper` and install them specifically (and use `FTDDIR`).

I hope that helps clear things up. The original logic was done the way it was so that we can support all the projects/devices in the same way and this shouldn't really change. I'm fine with adding some flexibility but it should be generic enough so that if other projects/devices can use this feature all they need to do is list extra dtb's in `uboot_helper`. Adding further platform specific code paths is a step in the wrong direction (IMO).

**Comment by vpeter4** [comment]
If copying dtb files using wildcard is problematic then just add all the files in uboot_helper file and copy them using loop. It will work if one file is set or more.

And this uboot_helper script should be separated to program and data: data should be separated for a project and reside in project folder. Or data should even be in project or devices option file.

**Comment by chewitt** [comment]
I'm quite aware of what the install/release/mkimage scripts do and I've committed [sic] worse crimes in the Amlogic dirs (not a good precedent to follow I know). I'd prefer to get this out of the way to let things move forwards, and make uboot_system prettier in a follow-up PR.

File: projects/NXP/devices/iMX6/bootloader/canupdate.sh [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/bin/sh
```
Comment (by-gnome) [inline projects/NXP/devices/iMX6/bootloader/canupdate.sh:1]: Then  [usr/share/bootloader/canupdate.sh](https://github.com/LibreELEC/LibreELEC.tv/blob/master/projects/NXP/devices/iMX6/filesystem/usr/share/bootloader/canupdate.sh) should be dropped?

Comment (chewitt) [reply projects/NXP/devices/iMX6/bootloader/canupdate.sh:1]: I'm not sure what you're trying to point out?

Comment (chewitt) [reply projects/NXP/devices/iMX6/bootloader/canupdate.sh:1]: Yes, it can be dropped. I'll add that later.

**Comment by etrax79** [comment]
I had a look to the update process and (as expected) it did work on my cubox-i2eXw, but maybe there is one new issue, which was not adressed yet:
'make image' with UBOOT_SYSTEM results in an update .tar archive also containing the board specific u-boot.img and SPL file.
So doing an update based on this .tar it will update KERNEL, SYSTEM, dtb(s) and also u-boot.img and SPL.
Doing the same without the UBOOT_SYSTEM, these two files are not part of the update archive and will not be replaced.
I would guess that at a certain point this will result in a version conflict, using an older u-boot.img with newer dtb's?

**Comment by chewitt** [comment]
I wouldn't expect issues from this as the delta between u-boot and kernel in an older (mature state) SoC like iMX6 isn't going to be that large over time. As long as it boots to load the kernel, the kernel takes over. In the ARM world it's not uncommon to have old vendor u-boot(s) installed and the latest kernel.

**Comment by by-gnome** [comment]
> @by-gnome @etrax79 please test the images here https://chewitt.libreelec.tv/testing/9.80/ as I don't have iMX6 h/w

Works for Wandboards.

**Comment by vpeter4** [comment]
> > @by-gnome @etrax79 please test the images here https://chewitt.libreelec.tv/testing/9.80/ as I don't have iMX6 h/w
> 
> Works for Wandboards.

Also for CuBox i4P.

**Comment by chewitt** [comment]
Testing by @etrax79 shows the fancy `canupdate.sh` is not required so i've simplified it. Older OE/LE images do not support canupdate checks so they will unpack the tar file and then fail on reboot due to wrong boot files. More recent LE 8.0+ images support the canupdate check but the older kernel (which performs the update) probably does not support zstd which is used to compress the newer image so the upgrade will fail to extract the canupdate file.

**Review by CvH — APPROVED** [review]
if changes are needed it could get PRed later, at least this gets the ball rolling

**Pull request merged**